## Análise: Otimização de preços e mix de produtos de uma empresa fictícia

### Criando o modelo matemático para otimização
#### Parâmetros
- Ai = Número máximo de smartphones modelo tipo i para vender este mês, onde i pertence ao conjunto {Lua1, Lua2}
- Bi = Preço de venda de smartphones modelo tipo i, onde i pertence ao conjunto {Lua1, Lua2}
- Ci = Preço de custo das peças componentes para smartphones modelo tipo i, onde i pertence ao conjunto {Lua1, Lua2}
- Di = Custo de mão de obra de montagem por hora de smartphones modelo tipo i, onde i pertence ao conjunto {Lua1, Lua2}
- Ei = Custo de mão de obra de teste por hora de smartphones modelo tipo i, onde i pertence ao conjunto {Lua1, Lua2}
- F = Número máximo de horas de trabalho de montagem
- G = Número máximo de horas de trabalho de teste
- Hm,i = Horas de montagem necessárias para construir cada modelo de smartphone tipo i, onde i pertence ao conjunto {Lua1, Lua2}
- Ht,i = Horas de teste necessárias para testar cada modelo de smartphone tipo i, onde i pertence ao conjunto {Lua1, Lua2}

#### Variável objetivo
- Xi = Número de smartphones modelo tipo i a produzir este mês, onde i pertence ao conjunto {Lua1, Lua2}

#### Função objetivo
LucroTotal = ∑(Xi * Bi) - ∑(Xi * Hm,i * Di) - ∑(Xi * Ht,i * Ei) - ∑(Xi * Ci)

#### A função objetivo indica:
LucroTotal = Faturamento - Custo para Produção(Montagem) - Custo para Produção(Teste) - Custo das Peças(Componentes)

#### Restrições
- O número de smartphones modelo tipo i a serem produzidos não pode ser negativo, ou seja, Xi >= 0, onde i pertence ao conjunto {Lua1, Lua2}.
- O número total de horas de montagem não pode ser maior que o número máximo de horas de mão de obra de montagem disponíveis.
- O número total de horas de teste não pode ser maior do que o máximo de horas de mão de obra de teste disponíveis.
- O número de smartphones modelo tipo i a serem produzidos não pode ser maior do que o número máximo de smartphones modelo tipo i a serem vendidos neste mês, onde i pertence ao conjunto {Lua1, Lua2}.

### Instalando e Carregando os pacotes

In [1]:
!pip install -q pulp

In [2]:
# Imports 
# Pacote de modelagem para otimização linear
from pulp import *

### Implementando o modelo matemático

In [3]:
# Número máximo de smartphones para vender este mês
A_Lua1 = 600
A_Lua2 = 1200

In [4]:
# Preço de venda dos smartphones
B_Lua1 = 300
B_Lua2 = 450

In [5]:
# Preço de custo dos smartphones
C_Lua1 = 150
C_Lua2 = 225

In [6]:
# Valor da hora para montagem
D_Lua1 = 11
D_Lua2 = 11

In [7]:
# Valor da hora para efetuar teste
E_Lua1 = 15
E_Lua2 = 15

In [8]:
# Número máximo de horas para montagem 
F = 10000

In [9]:
# Número máximo de horas para teste
G = 3000

In [10]:
# Horas de montagem necessárias para cada modelo de smartphone
Hm_Lua1 = 5
Hm_Lua2 = 6

In [11]:
# Horas de teste necessárias para testar cada modelo de smartphone
Ht_Lua1 = 1
Ht_Lua2 = 2

### Criação da Variável Para o Problema de Otimização

In [12]:
# Variável para o problema
problema = LpProblem("MixProdutos", LpMaximize)
problema

MixProdutos:
MAXIMIZE
None
VARIABLES

In [13]:
help(LpVariable)

Help on class LpVariable in module pulp.pulp:

class LpVariable(LpElement)
 |  LpVariable(name, lowBound=None, upBound=None, cat='Continuous', e=None)
 |  
 |  This class models an LP Variable with the specified associated parameters
 |  
 |  :param name: The name of the variable used in the output .lp file
 |  :param lowBound: The lower bound on this variable's range.
 |      Default is negative infinity
 |  :param upBound: The upper bound on this variable's range.
 |      Default is positive infinity
 |  :param cat: The category this variable is in, Integer, Binary or
 |      Continuous(default)
 |  :param e: Used for column based modelling: relates to the variable's
 |      existence in the objective function and constraints
 |  
 |  Method resolution order:
 |      LpVariable
 |      LpElement
 |      builtins.object
 |  
 |  Methods defined here:
 |  
 |  __init__(self, name, lowBound=None, upBound=None, cat='Continuous', e=None)
 |      Initialize self.  See help(type(self)) for 

In [14]:
# Define as variáveis
x_Lua1 = LpVariable("Unidades Lua1", 0, None, LpInteger)
x_Lua2 = LpVariable("Unidades Lua2", 0, None, LpInteger)

In [15]:
print(x_Lua1)
print(x_Lua2)

Unidades_Lua1
Unidades_Lua2


### Implementação da função objetivo

In [16]:
# Faturamento
faturamento = (x_Lua1 * B_Lua1) + (x_Lua2 * B_Lua2)
faturamento

300*Unidades_Lua1 + 450*Unidades_Lua2 + 0

In [17]:
# Custo Montagem
custo_montagem = (x_Lua1 * Hm_Lua1 * D_Lua1) + (x_Lua2 * Hm_Lua2 * D_Lua2)
custo_montagem

55*Unidades_Lua1 + 66*Unidades_Lua2 + 0

In [18]:
# Custo testes
custo_teste = (x_Lua1 * Ht_Lua1 * E_Lua1) + (x_Lua2 * Ht_Lua2 * E_Lua2)
custo_teste

15*Unidades_Lua1 + 30*Unidades_Lua2 + 0

In [19]:
# Custo das peças
custo_pecas = (x_Lua1 * C_Lua1) + (x_Lua2 * C_Lua2)
custo_pecas

150*Unidades_Lua1 + 225*Unidades_Lua2 + 0

In [20]:
# Substituindo as variáveis na função objetivo
problema += faturamento - custo_montagem - custo_teste - custo_pecas
problema

MixProdutos:
MAXIMIZE
80*Unidades_Lua1 + 129*Unidades_Lua2 + 0
VARIABLES
0 <= Unidades_Lua1 Integer
0 <= Unidades_Lua2 Integer

### Restrições

In [21]:
# O número máximo de horas de montagem disponíveis.
problema += (x_Lua1 * Hm_Lua1) + (x_Lua2 * Hm_Lua2) <= F,"Número Máximo de Horas de Montagem"

In [22]:
# O número máximo de horas de teste disponíveis.
problema += (x_Lua1 * Ht_Lua1) + (x_Lua2 * Ht_Lua2) <= G, "Número Máximo de Horas de Teste"

In [23]:
# Produção menor ou igual a demanda pelo modelo Lua1
problema += x_Lua1 <= A_Lua1,"Produção menor ou igual a demanda pelo modelo Lua1"

In [24]:
# Produção menor ou igual a demanda pelo modelo Lua2
problema += x_Lua2 <= A_Lua2,"Produção menor ou igual a demanda pelo modelo Lua2"

In [25]:
# Problema final
problema

MixProdutos:
MAXIMIZE
80*Unidades_Lua1 + 129*Unidades_Lua2 + 0
SUBJECT TO
Número_Máximo_de_Horas_de_Montagem: 5 Unidades_Lua1 + 6 Unidades_Lua2 <= 10000

Número_Máximo_de_Horas_de_Teste: Unidades_Lua1 + 2 Unidades_Lua2 <= 3000

Produção_menor_ou_igual_a_demanda_pelo_modelo_Lua1: Unidades_Lua1 <= 600

Produção_menor_ou_igual_a_demanda_pelo_modelo_Lua2: Unidades_Lua2 <= 1200

VARIABLES
0 <= Unidades_Lua1 Integer
0 <= Unidades_Lua2 Integer

In [26]:
### Resolvendo o problema de otimização
problema.solve()

1

In [27]:
# Lucro Maximizado
print("Lucro Maximizado: ", value(problema.objective))

Lucro Maximizado:  199600.0


In [28]:
# Número de Unidades para Produzir - Modelo 1
print("Número de Unidades do Modelo 1 a Produzir: ", problema.variables()[0].varValue)

Número de Unidades do Modelo 1 a Produzir:  560.0


In [29]:
# Número de Unidades para Produzir - Modelo 2
print("Número de Unidades do Modelo 2 a Produzir: ", problema.variables()[1].varValue)

Número de Unidades do Modelo 2 a Produzir:  1200.0


### Conclusão
A empresa Lua Smart Tech deve produzir 560 unidades do modelo Lua1 e 1200 unidades do modelo Lua2 para atingir o lucro máximo de R$199.600 .